src/common/download.h

```diff
@@ -32,5 +32,11 @@
 
 namespace tools
 {
-  bool download(const std::string &path, const std::string &url);
+  typedef void *download_async_handle;
```

Comment:
In this context, `using` and `typedef` are identical. `using` also allows templates, like in my `span` class:
```c++
template<typename T>
using view = span<typename std::add_const<T>::type>;
```
Its primarily helpful because the C++ template "language" requires nested calls inside of structs. If you want to access a type, `typename foo::type` is verbose, and can be "aliased" once:
```c++
template<typename T>
using add_const_t = typename std::add_const<T>::type; // available in C++14

template<typename T>
using view = span<add_const_t<T>>;
```

----------------------------

I suggested `std::unique_ptr<>` to initiate memory cleanup. So the background thread would complete, but the memory would be active until the caller was done with the handle. If you use `std::shared_ptr<>` then only a forward declare is necessary:

```c++
struct foo;
std::shared_ptr<foo> download_async(...)
```
because `std::shared_ptr` uses additional memory for storing the deleter. I would use a `std::shared_ptr<>` if you wanted detaching behavior - the caller and thread stack would each have a `shared_ptr<>` to the same object. Detaching without a ref count is probably a headache, but I am not really sure.